### PR TITLE
Add XONSH_HISTORY_IGNORE_REGEX support

### PR DESCRIPTION
**Pull request by bcr**
This is a fix for #3279 using the strategy proposed by @anki-code with some slight modifications.

I will add unit tests and proposed changelog text, I just want to make sure this is generally the right direction. Works on my machine with the JSON implementation 😆 

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

### CODE REVIEW

File: xonsh/history/json.py [hunk 1]
```diff
@@ -445,6 +446,12 @@ def __init__(
             if save_cwd is not None
             else XSH.env.get("XONSH_HISTORY_SAVE_CWD", True)
         )
+        ignore_regex = XSH.env.get("XONSH_HISTORY_IGNORE_REGEX")
```
Comment (jnoortheen) [inline xonsh/history/json.py:449]: this better be moved to the historybackend base class.

File: xonsh/history/base.py [hunk 2]
```diff
@@ -73,6 +73,7 @@ def __init__(self, sessionid=None, **kwargs):
         self.hist_size = None
         self.hist_units = None
         self.remember_history = True
+        self.should_append = lambda cmd: True
```
Comment (jnoortheen) [inline xonsh/history/base.py:76]: this can be a regular method. It will be more explicit and easier to extend if needed.

Comment (jnoortheen) [reply xonsh/history/base.py:76]: I also suggest naming the method something like, `is_ignored(cmd) -> bool`

File: xonsh/history/json.py (hunk 1 above)
Comment (bcr) [reply xonsh/history/json.py:449]: The discussion in #3279 was explicit about not putting it in the backend. I personally don't have any stake in this, I'll put it wherever you want, but if you and @anki-code don't agree, I'm not sure what to do about that.

File: xonsh/history/base.py [hunk 3]
```diff
@@ -164,3 +165,20 @@ def clear(self):
...
+        Call this in your append method.
+
+        Parameters
+        ----------
+        cmd: The prospective item to append.
+
+        Returns
+        -------
+        True if the item should be appended, False if not.
+        """
```
Comment (jnoortheen) [inline xonsh/history/base.py:180]: please check the [numpy-docstring](https://numpydoc.readthedocs.io/en/latest/format.html#parameters) format for parameters. The description should appear in the next line with indent

File: xonsh/history/json.py [hunk 4]
```diff
@@ -446,12 +446,7 @@ def __init__(
             if save_cwd is not None
             else XSH.env.get("XONSH_HISTORY_SAVE_CWD", True)
         )
-        ignore_regex = XSH.env.get("XONSH_HISTORY_IGNORE_REGEX")
-        self.is_ignored = (
-            (lambda cmd: re.match(ignore_regex, cmd["inp"]) is not None)
-            if ignore_regex is not None
-            else self.is_ignored
-        )
+        self.history_ignore_regex = XSH.env.get("XONSH_HISTORY_IGNORE_REGEX")
```
Comment (jnoortheen) [inline xonsh/history/json.py:448]: I think it is better to move this to the base class as well. We don't need to repeat unless there is any specialization

File: xonsh/environ.py [hunk 5]
```diff
@@ -1580,6 +1580,7 @@ class PromptHistorySetting(Xettings):
         "Save current working directory to the history.",
         doc_default="True",
     )
+    # !!! XONSH_HISTORY_IGNORE_REGEX goes here?
```
Comment (jnoortheen) [inline xonsh/environ.py:1583]: yes you can add it here

File: xonsh/history/json.py [hunk 6]
```diff
@@ -467,7 +468,7 @@ def append(self, cmd):
         hf : JsonHistoryFlusher or None
             The thread that was spawned to flush history
         """
-        if not self.remember_history:
+        if not self.remember_history or self.is_ignored(cmd):
             return
```
Comment (jnoortheen) [inline xonsh/history/json.py:472]: it is better to put first not expression inside a bracket

File: xonsh/history/json.py (hunk 4 above)
Comment (bcr) [reply xonsh/history/json.py:448]: `History` is not currently aware of `XSH`. Is it OK to introduce this? I'm cool with it if you are. @anki-code I'm defending your honor, you got an opinion?

Do you want an intermediate class from which both the JSON and SQLite derived, and we leave History ignorant of `XSH`?

File: tests/test_history_dummy.py [hunk 7]
```diff
@@ -10,3 +10,19 @@
 def test_construct_history_str(xession, backend):
     xession.env["XONSH_HISTORY_BACKEND"] = backend
     assert isinstance(construct_history(), DummyHistory)
+
+
+@pytest.mark.parametrize("backend", ["dummy", DummyHistory, DummyHistory()])
+def test_ignore_regex_invalid(xession, backend):
+    xession.env["XONSH_HISTORY_BACKEND"] = backend
+    xession.env["XONSH_IGNORE_REGEX"] = "**"
```
Comment (bcr) [inline tests/test_history_dummy.py:18]: I can't figure out how to inject `XONSH_IGNORE_REGEX` -- I don't think that `XSH.env` in the `History` class can see it.

Comment (jnoortheen) [reply tests/test_history_dummy.py:17]: setting them here is fine. It should be reflected during test run. You can see similar usage by other tests

File: xonsh/history/dummy.py [hunk 8]
```diff
@@ -7,6 +7,9 @@
 class DummyHistory(History):
     """A dummy implement of history backend."""
 
+    def __init__(self, **meta):
+        super().__init__(meta)
```
Comment (jnoortheen) [inline xonsh/history/dummy.py:11]: this should pass the '**kwargs'?

File: xonsh/history/base.py [hunk 9]
```diff
@@ -163,3 +169,35 @@ def clear(self):
...
         """
         pass
+
+    def validate_ignore_regex(self):
+        """Validates self.history_ignore_regex to make sure it's a valid
+        regex. If invalid, sets it to None and outputs a warning."""
+        if self.history_ignore_regex is not None:
+            try:
+                re.compile(self.history_ignore_regex)
+            except re.error:
```
Comment (jnoortheen) [inline xonsh/history/base.py:179]: I think we can store the compile result lazily. Like a `cached_property`

File: tests/test_history_dummy.py (hunk 7 above)
Comment (bcr) [reply tests/test_history_dummy.py:17]: In this case empirically it did not work. I'll hunt around and also make sure I didn't spell something wrong.

File: xonsh/history/dummy.py (hunk 8 above)
Comment (bcr) [reply xonsh/history/dummy.py:11]: I cribbed it from `JsonHistory.__init__`. I did forget the `**`, I'll fix it.

File: xonsh/history/base.py (hunk 9 above)
Comment (bcr) [reply xonsh/history/base.py:179]: This is just meant to validate the expression, but the point is taken that I can compile it once and use `match` on the compiled expression.

File: tests/test_history_dummy.py (hunk 7 above)
Comment (bcr) [reply tests/test_history_dummy.py:17]: `XONSH_HISTORY_IGNORE_REGEX` != `XONSH_IGNORE_REGEX` 🤦

**Comment by bcr** [comment]
I don't have any further changes that I'm thinking about, I think we've covered all of the comments so far, there are unit tests and docs in place, is this in good enough shape? Happy to hear any change suggestions. @jnoortheen @anki-code

**Comment by jnoortheen** [comment]
@bcr this looks good. One small nit though, still the regex is not compiled. It would be faster when compiled and matched. The easier way would be to add a `cached_property` like

```py
# this is just a psuedo-code, feel free to use what you think the best and concise

import functools

class BaseHistory():
  @functools.cached_property
  def ignore_regex(self):
      regex = XSH.env.get("XONSH_HISTORY_IGNORE_REGEX")
      if regex:
        return re.compile(regex)

  def is_ignored(self, cmd):
    return self.ignore_regex and self.ignore_regex.match(cmd)
  ...
```

**Comment by bcr** [comment]
Let me take a whack at this. @anki-code had it as a string, so I was honoring that. Understood that having it pre-compiled will be faster.

**Comment by bcr** [comment]
@jnoortheen it is done. May our unit tests be mighty and save us from folly.

File: xonsh/history/base.py [hunk 10]
```diff
@@ -163,3 +169,36 @@ def clear(self):
         memory.
         """
         pass
+
+    @functools.cached_property
+    def ignore_regex(self):
+        self.history_ignore_regex = None
```
Comment (jnoortheen) [inline xonsh/history/base.py:175]: you don't have to store it in an attribute. that will be handled by the decorator

**Review by jnoortheen — APPROVED** [review]

**Pull request merged**